openhtf/io/station_api.py

```diff
@@ -0,0 +1,501 @@
+
+  def get_history_after(self, test_uid, start_time_millis):
+    """Get a list of pickled TestRecords for test_uid from the History."""
+    # TODO(madsci): We really should pull attachments out of band here.
+    return [saxutils.escape(pickle.dumps(test_record))
+            for test_record in history.for_test_uid(
+                test_uid, start_after_millis=start_time_millis)]
+
+
+class Server(threading.Thread):
+
+  daemon = True
+
+  def __init__(self):
+    super(Server, self).__init__()
+
+    self.station_api_server = None
+    self.multicast_listener = None
+    self.last_activity_time_millis = 0
+
+  def multicast_response(self, message):
+    if message != conf.station_discovery_string:
+      _LOG.debug('Received unexpected traffic on discovery socket: %s', message)
+    else:
+      return json.dumps({
+          'station_id': conf.station_id,
+          'station_api_bind_address': conf.station_api_bind_address,
+          'station_api_port': conf.station_api_port,
+          'last_activity_time_millis': self.last_activity_time_millis,
+      })
+
+  def run(self):
+    if int(conf.station_api_port):
+      self.station_api_server = SimpleXMLRPCServer.SimpleXMLRPCServer(
+          (conf.station_api_bind_address, int(conf.station_api_port)),
+          allow_none=True,
+          logRequests=logging.getLogger().level <= logging.DEBUG)
+      self.station_api_server.register_instance(STATION_API)
+      self.station_api_server.register_introspection_functions()
+      self.station_api_server.register_multicall_functions()
+
+      # Discovery is useless if station_api is disabled, so we don't ever start
+      # a MulticastListener if station_api_port isn't set, even if
+      # enable_station_discovery is set.
+      if conf.enable_station_discovery:
+        self.multicast_listener = multicast.MulticastListener(
+            self.multicast_response,  **MULTICAST_KWARGS())
+        _LOG.debug(
+            'Listening for multicast discovery at %s:%s',
+            self.multicast_listener.address, self.multicast_listener.port)
+        self.multicast_listener.start()
+
+      # server_forever() doesn't return until we call stop()
+      _LOG.debug(
+          'Starting station_api server on port %s', conf.station_api_port)
+      self.station_api_server.serve_forever()
+      _LOG.debug('station_api server exiting.')
+    else:
+      _LOG.debug('Started station_api, but station_api_port disabled, bailing.')
+
+  def stop(self):
+    try:
+      try:
+        if self.multicast_listener:
+          self.multicast_listener.Stop()
+      finally:
+        if self.station_api_server:
+          self.station_api_server.shutdown()
+    except Exception:
+      _LOG.debug(
+          'Exception shutting down %s', type(self), exc_info=True)
+
+
+# Singleton instances.
+STATION_API = StationApi()
+API_SERVER = None
+
+def start_server():
+  # TODO(madsci): Blech, fix this.
```

Comment:
any suggestions on what to do instead, here?  or is this just ugly-but-the-best-way?